contracts/liquidator/OnChainLiquidator.sol

```diff
@@ -0,0 +1,539 @@
+        }
+
+        emit Swap(swapToken, baseToken, swapAmount, amountOut, Exchange.SushiSwap, 3000);
+
+        return amountOut;
+    }
+
+    function swapViaBalancer(address comet, address asset, uint256 amountOutMin, PoolConfig memory poolConfig) internal returns (uint256) {
+        uint256 swapAmount = ERC20(asset).balanceOf(address(this));
+        // Safety check, make sure residue balance in protocol is ignored
+        if (swapAmount == 0) return 0;
+
+        address swapToken = asset;
+
+        address baseToken = CometInterface(comet).baseToken();
+
+        // XXX move?
+        address BALANCER_VAULT = 0xBA12222222228d8Ba445958a75a0704d566BF2C8;
+
+        TransferHelper.safeApprove(asset, address(BALANCER_VAULT), swapAmount);
+
+        int256[] memory limits = new int256[](2);
+        limits[0] = type(int256).max;
+        limits[1] = type(int256).max;
+
+        IAsset[] memory assets = new IAsset[](2);
+        assets[0] = IAsset(asset);
+        assets[1] = IAsset(baseToken);
+
+        IVault.BatchSwapStep[] memory steps = new IVault.BatchSwapStep[](1);
+        steps[0] = IVault.BatchSwapStep({
+            poolId: poolConfig.balancerPoolId,
+            assetInIndex: 0,
+            assetOutIndex: 1,
+            amount: swapAmount,
+            userData: bytes("")
+        });
+
+        int256[] memory assetDeltas = IVault(BALANCER_VAULT).batchSwap(
+            IVault.SwapKind.GIVEN_IN,
+            steps,
+            assets,
+            IVault.FundManagement({
+                sender: address(this),
+                fromInternalBalance: false,
+                recipient: payable(this),
+                toInternalBalance: false
+            }),
+            limits, // limits
+            block.timestamp
+        );
+
+        int256 signedAmountOut = -assetDeltas[assetDeltas.length - 1];
+
+        if (signedAmountOut < 0) {
+            // XXX custom error
+            revert("signedAmoutOut cannot be negative");
+        }
+
+        uint256 amountOut = uint256(signedAmountOut);
+
+        if (amountOut < amountOutMin) {
+            revert InsufficientAmountOut(swapToken, baseToken, swapAmount, amountOut, amountOutMin, Exchange.Balancer, 0);
+        }
+
+        return amountOut;
+    }
+
+    function swapViaCurve(address comet, address asset, uint256 amountOutMin, PoolConfig memory poolConfig) internal returns (uint256) {
+        uint256 swapAmount = ERC20(asset).balanceOf(address(this));
+        // Safety check, make sure residue balance in protocol is ignored
+        if (swapAmount == 0) return 0;
+
+        address swapToken = asset;
+        address baseToken = CometInterface(comet).baseToken();
+
+        address CURVE_REGISTRY = 0x90E00ACe148ca3b23Ac1bC8C240C2a7Dd9c2d7f5;
+        address ST_ETH = 0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84;
+        address WST_ETH = 0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0;
+        address ALL_EES = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
```

Comment:
- [ ] move these addresses to the contract level
- [ ] rename `ALL_EES` to something less stupid